### PR TITLE
[mlir] Use find_package() to locate MLIR static library

### PR DESCRIPTION
**Pull request by zhanglx13**
The latest version of MLIR exports the static library that can be found by `find_package()`. The changes made on the MLIR side can be found [here](https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/569).

New names (static library related):
- MLIR project name: rocMLIR
- MLIR package name: rocMLIR
- MLIR static library name: rockCompiler

The corresponding changes on the MIOpen side are:
- Use `find_package(rocMLIR 1.0.0)` to find the `rocMLIR` package of version 1.0.0. This is the first version of MLIR that exports the static library.
- For backward compatibility, use `find_library()` to locate the library in the old way if `find_package()` does not find the package.
- Add `rocMLIR::rockCompiler` to the link library of target `MIOpen`.
- Update the commit id of MLIR in `requirements.txt`.

### CODE REVIEW

**Comment by jerryyin** [comment]
@zhanglx13 Could you update https://github.com/ROCmSoftwarePlatform/MIOpen/blob/develop/requirements.txt so that on MIOpen end it can pull in latest commit and run through MIOpen regressions?

**Comment by jerryyin** [comment]
@zhanglx13 FYI this is a ROCm 5.4 feature so can only merge after MIOpen branches.

Could you take a look at test failures and try to work through? Just so that when MIOpen finished branching we are good to go.

**Comment by krzysz00** [comment]
I thought MIOpen had branched already? In any case, this is still blocking deploying my changes because of the order the changes are going into MLIR.

File: driver/CMakeLists.txt
Comment (JehandadKhan) [inline driver/CMakeLists.txt:31]: If the name of the package is `miir` then why are the variable names following the old convention?

Comment (zhanglx13) [reply driver/CMakeLists.txt:31]: I was thinking `MLIRMIOpen` is the library name. 
And our team is in the middle of name changing. The `Miir` can also be changed in the near future.

Comment (jerryyin) [reply driver/CMakeLists.txt:31]: That's right. We plan rename all those in ROCm 5.4 time frame (including the MLIR project name!). It's going to me `rocMLIR` and `rock` dialect. Project names referenced here will change.

Speaking of which @zhanglx13 maybe we want to talk to @sjw36  and determine the name so the package name can go in as the ultimate form. We can amend `miir` header name in another PR?

Comment (zhanglx13) [reply driver/CMakeLists.txt:31]: @jerryyin Sure. Changing package and library names can be done any time and it should not affect developer on the MIOpen team as long as wedo not change the API name. 
Currently I am figuring out the test failures, which has something to do with the dependencies. After everything is resolved, this PR should be good to go.

Comment (jerryyin) [reply driver/CMakeLists.txt:31]: Sounds good to me. Just a heads up though, the name change will have to happen inside a single release otherwise we'll be facing with backward compatibility breakages. Let's coordinate it after this PR.

File: src/CMakeLists.txt
Comment (pfultz2) [inline src/CMakeLists.txt:639]: The includes should come from the cmake targets, not from cmake variables.

Comment (zhanglx13) [reply src/CMakeLists.txt:639]: The variable `MLIRMIOpen_INCLUDE_DIR` is set in `MiirConfig.cmake`, which is auto-generated when building the MLIR project ([see code here](https://github.com/ROCmSoftwarePlatform/llvm-project-mlir/blob/484f2897d1a4c7b54802cc83ab2d04bec41b8ed8/mlir/tools/mlir-miopen-lib/cmake/MiirConfig.cmake.in#L4)) and loaded first by `find_package(Miir)`. Is this what you are suggesting @pfultz2 ?

Comment (pfultz2) [reply src/CMakeLists.txt:639]: We shouldn't use the variables such as `MLIRMIOpen_INCLUDE_DIR` as they dont contain all the usage requirements. We should just use the cmake targets, so `target_link_libraries(MIOpen PRIVATE Miir::MLIRMIOpen)` should enough. If thats not the case, then its a bug in mlir which needs to be fixed, and we shouldn't need a workaround.

File: driver/CMakeLists.txt
Comment (pfultz2) [reply driver/CMakeLists.txt:31]: This should not be added here. Instead you should do `target_internal_library(MIOpen Miir::MLIRMIOpen)` so the usage requirements of mlir are propagated to the driver and the tests.

File: src/CMakeLists.txt
Comment (zhanglx13) [reply src/CMakeLists.txt:639]: The new MLIR installs the static library as an exported target. Therefore, it needs to tell MIOpen targets the location of the header file `Miir.h`.

File: driver/CMakeLists.txt
Comment (zhanglx13) [reply driver/CMakeLists.txt]: removed since MIOpenDriver is not including Miir.h

**Comment by zhanglx13** [comment]
@junliume Do you know how to rebuild the docker image? For now, Jenkins let me set the parameter `DOCKER_IMAGE_OVERRIDE` but it does not rebuild the image if a new name is given.

File: CMakeLists.txt
Comment (shurale-nkn) [inline CMakeLists.txt:327]: find_package info: Regardless of the mode used, a \<PackageName\>_FOUND variable will be set to indicate whether the package was found.
\<PackageName\>_FOUND will be a more understandable solution

File: src/CMakeLists.txt
Comment (shurale-nkn) [reply src/CMakeLists.txt:639]: I think @pfultz2 wants to say that we shouldn't do this manually. If the package was assembled correctly, after finding it cmake can automatically transfer "public dependencies of MLIRMIOpen" to MIOpen if MLIR was part of 'target_link_libraries'.

Comment (zhanglx13) [reply src/CMakeLists.txt]: @pfultz2 @shurale-nkn Thank you for the explanation. I got the include_dir information embedded into the export target as its property so that the `target_link_libraries()` is the only needed by MIOpen to use libMLIRMIOpen.

**Comment by JehandadKhan** [comment]
> @junliume Do you know how to rebuild the docker image? For now, Jenkins let me set the parameter `DOCKER_IMAGE_OVERRIDE` but it does not rebuild the image if a new name is given.
> 
> The [last successful CI](http://micimaster.amd.com/blue/organizations/jenkins/MLLibs%2FMIOpen/detail/find-MLIR/28/pipeline/301#step-470-log-78) build shows that the compatibility function is working well
> 
> ```
> [2022-08-18T21:39:49.140Z] -- Could NOT find Miir (missing: Miir_DIR)
> [2022-08-18T21:39:49.140Z] -- Falling back to find library libMLIRMIOpen
> [2022-08-18T21:39:49.140Z] -- Build with library libMLIRMIOpen: /opt/rocm/lib/libMLIRMIOpen.a
> ```
> 
> And this also shows that the image used contains the old version of MLIR static library. To test the new exported library, a image needs to be built with the updated version of MLIR in the requirements.txt.

Hello @zhanglx13. You can rebuild the docker yourself with the docker build arguments that our CI echos, push it to either the internal or the external registry and then override the CI runs using the `DOCKER_IMAGE_OVERRIDE` parameter. Once your PR is merged we can update the image that the CI uses by default.

**Comment by zhanglx13** [comment]
The old MLIR is checked with [build 32](http://micimaster.amd.com/job/MLLibs/job/MIOpen/job/find-MLIR/32/).
The new MLIR is checked with [build 40](http://micimaster.amd.com/job/MLLibs/job/MIOpen/job/find-MLIR/40/).

**Review by jerryyin — APPROVED** [review]

**Review by shurale-nkn — DISMISSED** [review]
LGTM

File: CMakeLists.txt
Comment (atamazov) [inline CMakeLists.txt:327]: Please indicate the max ROCm version this piece of code maintains backward compatibility with. This would help us clean up the code in the future. For example, if this is 5.3, then we will be able to remove this piece of code as soon 5.3 support is dropped.

File: src/CMakeLists.txt
Comment (atamazov) [inline src/CMakeLists.txt:687]: Ditto.

**Review by atamazov — COMMENTED** [review]
Almost good to go!

File: CMakeLists.txt
Comment (jerryyin) [reply CMakeLists.txt]: Note: because of the backward compatibility requirement I'd prefer the package name to be renamed to the correct one instead of `Miir` right now. (If we do it in next release we'd have to carry this package name in future for backward compatibility purpose again)

**Comment by atamazov** [comment]
@zhanglx13 
> [zhanglx13](https://github.com/zhanglx13) [force-pushed](https://github.com/ROCmSoftwarePlatform/MIOpen/compare/62a716b6ed9c2e58758a6c3a093b4c6aea9a693b..dc4994c34d961079580b1df798025f390b9140f7) the find-MLIR branch from[ 62a716b ](https://github.com/ROCmSoftwarePlatform/MIOpen/commit/62a716b6ed9c2e58758a6c3a093b4c6aea9a693b)to[ dc4994c ](https://github.com/ROCmSoftwarePlatform/MIOpen/commit/dc4994c34d961079580b1df798025f390b9140f7)2 hours ago

Please do not **_force-push_**. This is NOT recommended due to several reasons. Force-pushing can be more or less safely used only if the branch has been published for review yet.

File: CMakeLists.txt
Comment (atamazov) [inline CMakeLists.txt:339]: [Recommendation] Normally we use indentation of 4 spaces, not 2.

**Comment by atamazov** [comment]
> @zhanglx13
> 
> > [zhanglx13](https://github.com/zhanglx13) [force-pushed](https://github.com/ROCmSoftwarePlatform/MIOpen/compare/62a716b6ed9c2e58758a6c3a093b4c6aea9a693b..dc4994c34d961079580b1df798025f390b9140f7) the find-MLIR branch from[ 62a716b ](https://github.com/ROCmSoftwarePlatform/MIOpen/commit/62a716b6ed9c2e58758a6c3a093b4c6aea9a693b)to[ dc4994c ](https://github.com/ROCmSoftwarePlatform/MIOpen/commit/dc4994c34d961079580b1df798025f390b9140f7)2 hours ago
> 
> Please do not **_force-push_**. This is NOT recommended due to several reasons. Force-pushing can be more or less safely used only if the branch has been published for review yet.

For example, this makes previous code reviews invisible on the "Code changes" tab (due to overwritten commit hashes, I guess).

**Review by atamazov — APPROVED** [review]
LGTM unless you would like to fix formatting

File: CMakeLists.txt
Comment (atamazov) [reply CMakeLists.txt]: @jerryyin Very good!

**Comment by zhanglx13** [comment]
> > @zhanglx13
> > > [zhanglx13](https://github.com/zhanglx13) [force-pushed](https://github.com/ROCmSoftwarePlatform/MIOpen/compare/62a716b6ed9c2e58758a6c3a093b4c6aea9a693b..dc4994c34d961079580b1df798025f390b9140f7) the find-MLIR branch from[ 62a716b ](https://github.com/ROCmSoftwarePlatform/MIOpen/commit/62a716b6ed9c2e58758a6c3a093b4c6aea9a693b)to[ dc4994c ](https://github.com/ROCmSoftwarePlatform/MIOpen/commit/dc4994c34d961079580b1df798025f390b9140f7)2 hours ago
> > 
> > 
> > Please do not **_force-push_**. This is NOT recommended due to several reasons. Force-pushing can be more or less safely used only if the branch has been published for review yet.
> 
> For example, this makes previous code reviews invisible on the "Code changes" tab (due to overwritten commit hashes, I guess).

@atamazov Thank you for the notes. Sorry for the inconvenience.

I'll fix the indentation together with a new name for the package and library as soon as we settle the names.

**Review by atamazov — DISMISSED** [review]
LGTM

**Review by jerryyin — DISMISSED** [review]

**Comment by zhanglx13** [comment]
@junliume Can you please help me take a look at the [docker image building failure](http://micimaster.amd.com/blue/organizations/jenkins/MLLibs%2FMIOpen/detail/find-MLIR/54/pipeline)? 
I was trying to build a new docker image to test the latest version of MLIR.

**Review by jerryyin — APPROVED** [review]

**Comment by zhanglx13** [comment]
Please do not merge this PR until the CI that builds a new docker image passes.

**Review by jerryyin — APPROVED** [review]

File: requirements.txt
Comment (jerryyin) [inline requirements.txt:7]: @JehandadKhan Looks like this `--offload-arch` has been blocking @zhanglx13 for weeks. It looks like a composable kernel repo's particular concept. Could you help explain what this is? (and why this is not caught by previous docker image builds?)

**Review by jerryyin — APPROVED** [review]

**Review by junliume — DISMISSED** [review]
@zhanglx13 have you updated the `rocm/miopen:ci` docker with the newer dependencies? If that docker is ready we can merge this PR asap.

**Comment by zhanglx13** [comment]
@junliume I haven't seen a successful build of the docker image with the new dependencies. Still trying to figure out the cause.

**Comment by zhanglx13** [comment]
There is a blocker on the CK side.
After https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/435 lands, we need to update the CK commit id in requirements and rebuild the docker image.

**Comment by zhanglx13** [comment]
Another issue caused by CK integration. See https://github.com/ROCmSoftwarePlatform/MIOpen/issues/1777 for details.

**Comment by zhanglx13** [comment]
Decision needs to be made by @junliume @JehandadKhan about CK integration. See #1777 for details.
If it is decided to bound with an old version of CK, then we need to revert `fd08663` and update CK's commit id in requirements.txt  to the chosen one.
If it is decided to keep up with the new CK, we can wait until all issues are resolved.

File: CMakeLists.txt
Comment (junliume) [inline CMakeLists.txt:226]: @zhanglx13 This `utility` is what I was mentioning about :)
Please try to build based on the latest docker image `rocm/miopen:ci_rocm530_ubuntu2004`

**Comment by zhanglx13** [comment]
Blocked by PR https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1761.

**Comment by krzysz00** [comment]
I'd just like to reiterate that the changes in rocMLIR that require this PR to land are already in our 5.4 release branch, and so therefore this PR must land on MIOpen's 5.4 release branch - or at least that's what I can tell of the situation

File: requirements.txt
Comment (junliume) [inline requirements.txt:5]: @jerryyin and @zhanglx13 can you point to an actual commit ID instead of a branch name? Unless `release/rocm-5.4` is a release tag already.

Comment (jerryyin) [reply requirements.txt:5]: 11f2e26f193d398cc67f00d9b8e5a58504a14092, [here](https://github.com/ROCmSoftwarePlatform/rocMLIR/commit/11f2e26f193d398cc67f00d9b8e5a58504a14092) is the last commit of ROCm 5.4 branch. You are right that we haven't tagged rocm-5.4.0 yet.

Comment (junliume) [reply requirements.txt:5]: @jerryyin MIOpen CI docker has been updated; could you update this line with commit hash? I would expect the next round of CI to pass :)

Comment (zhanglx13) [reply requirements.txt]: @junliume Which docker image is the latest one? `rocm/miopen:ci`?

Comment (junliume) [reply requirements.txt]: @zhanglx13 yes `rocm/miopen:ci` has the correct CK so this PR should go through; however, we need this PR in so that we can create a new docker image which installed rocMLIR@11f2e26f193d398cc67f00d9b8e5a58504a14092. ETA this Friday.

**Review by junliume — APPROVED** [review]

**Review by junliume — APPROVED** [review]

**Comment by zhanglx13** [comment]
@junliume [build#80](http://micimaster.amd.com/job/MLLibs/job/MIOpen/job/find-MLIR/80/) has all greens. However, `rocm/miopen:ci` does not use the rocm-5.4 version of MLIR. Therefore, I kicked off [build#81](http://micimaster.amd.com/blue/organizations/jenkins/MLLibs%2FMIOpen/detail/find-MLIR/81/pipeline), which builds a new docker image with the latest MLIR. I think we should be safe to merge this PR after we see all greens in [build#81](http://micimaster.amd.com/blue/organizations/jenkins/MLLibs%2FMIOpen/detail/find-MLIR/81/pipeline).

**Pull request merged**
